module-system/module-implementations/sov-blob-storage/src/capabilities.rs

```diff
@@ -16,36 +20,88 @@ impl<C: Context, Da: DaSpec> BlobSelector<Da> for BlobStorage<C, Da> {
             // Prevent double number of blobs being executed
-            return Ok(past_deferred
+            return Ok(expiring_deferred_blobs
                 .into_iter()
                 .map(Into::into)
                 .chain(current_blobs.into_iter().map(Into::into))
                 .collect());
         };
 
+        // If we reach this point, there is a preferred sequencer, so we need to handle cases 1 and 3.
+
+        // First, compute any "bonus blobs" requested
+        // to be processed early.
+        let num_bonus_blobs_requested = self
+            .deferred_blobs_requested_for_execution_next_slot
+            .get(working_set)
+            .unwrap_or_default();
+        self.deferred_blobs_requested_for_execution_next_slot
+            .set(&0, working_set);
+
+        let mut remaining_blobs_requested =
+            (num_bonus_blobs_requested as usize).saturating_sub(expiring_deferred_blobs.len());
+        let mut bonus_blobs: Vec<BlobRefOrOwned<Da::BlobTransaction>> =
+            Vec::with_capacity(remaining_blobs_requested);
+        let mut next_slot_to_check = slot_for_expiring_blobs + 1;
+        while remaining_blobs_requested > 0 && next_slot_to_check < current_slot {
+            let mut blobs_from_next_slot =
+                self.take_blobs_for_slot_height(next_slot_to_check, working_set);
+
+            // If the set of deferred blobs from the next slot in line contains more than the remainder needed to fill the request,
+            //  we split that group and save the unused portion back into state
+            if blobs_from_next_slot.len() > remaining_blobs_requested {
+                let blobs_to_save: Vec<<Da as DaSpec>::BlobTransaction> =
+                    blobs_from_next_slot.split_off(remaining_blobs_requested);
+                bonus_blobs.extend(blobs_from_next_slot.into_iter().map(Into::into));
+                self.store_blobs(
+                    next_slot_to_check,
+                    &blobs_to_save.iter().collect::<Vec<_>>(),
+                    working_set,
+                )?;
+                remaining_blobs_requested = 0;
+                break;
+            } else {
+                remaining_blobs_requested -= blobs_from_next_slot.len();
+                bonus_blobs.extend(blobs_from_next_slot.into_iter().map(Into::into));
+            }
+            next_slot_to_check += 1;
+        }
+
+        // Finally handle any new blobs which appeared on the DA layer in this slot
         let mut priority_blobs = Vec::new();
         let mut to_defer: Vec<&mut Da::BlobTransaction> = Vec::new();
-
         for blob in current_blobs {
+            // Blobs from the preferred sequencer get priority
             if blob.sender() == preferred_sequencer {
                 priority_blobs.push(blob);
             } else {
-                to_defer.push(blob);
+                // Other blobs get deferred unless the sequencer has requested otherwise
+                if remaining_blobs_requested != 0 {
```

Comment:
Nit: Let's change it to `remaining_blobs_requested > 0`. I know it is impossible to go negative here but I think `>` is more defensive.

Reply:
Good idea. Done.